### PR TITLE
WIP jinja2 convenience function

### PR DESCRIPTION
**Pull request by qduk**
This is just a WIP. I want to get your guys opinion on this approach for a jinja2 convenience function. I would still need to build tests, give examples, update documentation, add error handling and add some sort of validation for parameter inputs. The usage is how David Flores mentioned in #29. You would import the convenience function and then register it as a filter.  Thoughts?

### CODE REVIEW

**Comment by itdependsnetworks** [comment]
I was thinking something like
```
_JINJA_MAP = {
	"fqdn_to_ip": "netutils.dns.fqdn_to_ip",
	"is_fqdn_resolvable": "netutils.dns.is_fqdn_resolvable",
	etc...
}

def jinja2_convenience_function():
    result = {}
    for key, value in _JINJA_MAP.items():
        result[key] = import_module(value)

    return result
```


And example usage:
```
import jinja2

env = jinja2.Environment()
env.filters.extend(jinja2_convenience_function())

```

Then we can build a test that every function get's added to the _JINJA_MAP

**Comment by qduk** [comment]
That's a good idea. Didn't know if using one filter or splitting it up would be best. Hence why I wanted to run it by you guys. I'll update this PR with the refactor, proper tests, etc.

**Comment by qduk** [comment]
@itdependsnetworks thoughts? This presents each function as its own filter and dynamically finds the files needed. If a new folder was created the code would need updating but anything in the `root` or `config` directories will be included.

**Comment by itdependsnetworks** [comment]
Would be good to get someone else to chime in (@jeffkala ?? ) I still personally prefer the explicit mapping. I also think this should help for documentation as well as naming. As example, `get_usable_range` by itself is not as descriptive, outside of the context of being an ip check. Perhaps we would change the name to `get_usable_range_ip` or similar.

**Comment by jeffkala** [comment]
I like explicit mapping in this case.  I think its easier to read and understand the code that way.  Also would agree with Ken that the documentation would be easier.

**Comment by qduk** [comment]
@jeffkala @itdependsnetworks I revamped this. Let me know what you guys think.

File: netutils/utils.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+"""Utilities for the netutils library."""
+from importlib import import_module
+
+_JINJA2_FUNCTION_MAPPINGS = {
+    "asn_to_int": "asn.asn_to_int",
+    "config": {
```
Comment (itdependsnetworks) [inline netutils/utils.py:6]: I don't understand this nesting

File: docs/source/netutils/utilities/index.rst [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    from jinja2.loaders import FileSystemLoader, PackageLoader
+    from jinja2 import Environment, PackageLoader, select_autoescape
+    from netutils.utils import jinja2_convenience_function
+
+    env = Environment(
+        loader=FileSystemLoader("templates"),
+        autoescape=select_autoescape()
+    )
+
+    for function_name, function_object in jinja2_convenience_function().items():
```
Comment (itdependsnetworks) [inline docs/source/netutils/utilities/index.rst:37]: I said extend, but I meant to say update, does this work? env.filters.update(jinja2_convenience_function())

File: netutils/utils.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        Any: Return value depends on the function called.
+
+    """
+    if not jinja_mappings_dict:
+        jinja_mappings_dict = _JINJA2_FUNCTION_MAPPINGS
+
+    result = {}
+
+    for function_name, function_import_path in jinja_mappings_dict.items():
+        if isinstance(function_import_path, dict):
```
Comment (itdependsnetworks) [inline netutils/utils.py:70]: I don't think you need this, just spell it out in the mapping.

File: docs/source/netutils/utilities/index.rst [hunk 4]
```diff
@@ -0,0 +1,65 @@
+*******************
+Netutils Utilities
+*******************
+
+Netutils to Jinja2 Filters
+============================
+
+In an effort to simplify the process of adding netutil's functions to jinja2 as filters we have created a convenience function. Let me give you an example of one use case.
```
Comment (jeffkala) [inline docs/source/netutils/utilities/index.rst:8]:
```suggestion
In an effort to simplify the process of adding netutils' functions to jinja2 as filters we have created a convenience function. One use case.
```

File: docs/source/netutils/utilities/index.rst [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    IP Address + 200 = 192.168.0.201
+
+
+Netutils to Jinja2 Filters List
+======================================
+
+The below list shows what jinja2 filters are added when you add them using the process above. The keys of the dictionary are the names you would use to call the jinja2 filter.
+
+.. note::
+    Some names have been changed to from their function names to provide more context when used as jinja2 filters.
```
Comment (jeffkala) [inline docs/source/netutils/utilities/index.rst:57]:
```suggestion
    Some names have been changed from their python function names to provide more context when used as a jinja2 filter.
```

File: netutils/utils.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def jinja2_convenience_function(jinja_mappings_dict=None, parent_folder=None):
+    """Convenience function that allows netutils filter to be used easily with jinja2.
+
+    Returns:
+        Any: Return value depends on the function called.
+
+    """
+    if not jinja_mappings_dict:
```
Comment (jeffkala) [inline netutils/utils.py:64]: We want the ability to allow someone to specify their own mapping?  Not sure I understand why?

**Review by jeffkala — CHANGES_REQUESTED** [review]

File: docs/source/netutils/utilities/index.rst (hunk 2 above)
Comment (qduk) [reply docs/source/netutils/utilities/index.rst:37]: That does indeed work! Makes sense.

File: netutils/utils.py (hunk 1 above)
Comment (qduk) [reply netutils/utils.py]: I definitely made this more complicated than it needed to be. I blame it being Sunday when I did it and not know about `rstrip()`

File: netutils/utils.py (hunk 6 above)
Comment (qduk) [reply netutils/utils.py]: This has been refactored.

**Comment by qduk** [comment]
@jeffkala @itdependsnetworks I refactored this. I don't know why I originally made it so complicated. I blame it being Sunday when I did this and not knowing about `rstrip()`.

File: netutils/utils.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    "longest_prefix_match": "route.longest_prefix_match",
+    "vlanlist_to_config": "vlan.vlanlist_to_config",
+    "vlanconfig_to_list": "vlan.vlanconfig_to_list",
+}
+
+
+def jinja2_convenience_function():
+    """Convenience function that allows netutils filter to be used easily with jinja2.
+
+    Returns:
```
Comment (itdependsnetworks) [inline netutils/utils.py:58]: Returns a dict, no?

Comment (itdependsnetworks) [reply netutils/utils.py:58]: Can you add a testable example as well?

File: docs/source/netutils/utilities/index.rst [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+Netutils to Jinja2 Filters
+============================
+
+In an effort to simplify the process of adding netutils' functions to jinja2 as filters we have created a convenience function. One use case.
+Here is the current folder structure.
+
+.. code-block:: python
+
+    .
+    ├── p1.py
```
Comment (itdependsnetworks) [inline docs/source/netutils/utilities/index.rst:14]: Use logical names

Comment (qduk) [reply docs/source/netutils/utilities/index.rst]: Updated

**Pull request merged**